xds/internal/resolver/watch_service.go

```diff
@@ -0,0 +1,235 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package resolver
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"google.golang.org/grpc/internal/grpclog"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+)
+
+// serviceUpdate contains information received from LDS and RDS responses,
+// which is of interest to the registered service watcher.
+type serviceUpdate struct {
+	// Routes contain matchers+actions to route RPCs.
+	Routes []*xdsclient.Route
+}
+
+// watchService uses LDS and RDS to discover information about the provided
+// serviceName.
+//
+// Note that during race (e.g. an xDS response is received while the user is
+// calling cancel()), there's a small window where the callback can be called
+// after the watcher is canceled. The caller needs to handle this case.
+func watchService(c xdsClientInterface, serviceName string, cb func(serviceUpdate, error), logger *grpclog.PrefixLogger) (cancel func()) {
```

Comment:
I'm assuming most of this code is unchanged, just moved. Are there any changes that I should be looking at?

Reply:
No. They are just moved.